### PR TITLE
Add `ExtendedLayerSpec` with optional shared `encoding` and `projection`

### LINKED ISSUE #3344: Projections should propagate down
In the spec below, the projection should be used for both layers but as the image shows, the default projection is used. 

Reported by @jwoLondon

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.1.json",
  "width": 800,
  "height": 500,
  "data": {
    "url": "data/us-state-capitals.json"
  },
  "projection": {
    "type": "albersUsa"
  },
  "layer": [{
    "mark": "circle",
    "encoding": {
      "x": {
        "field": "lon",
        "type": "longitude"
      },
      "y": {
        "field": "lat",
        "type": "latitude"
      },
      "color": {"value": "orange"}
    }
  }, {
    "mark": "text",
    "encoding": {
      "x": {
        "field": "lon",
        "type": "longitude"
      },
      "y": {
        "field": "lat",
        "type": "latitude"
      },
      "text": {"field": "city", "type": "nominal"}
    }
  }]
}
```

screen shot 2018-02-01 at 00 33 10

### LINKED ISSUE #3384: Layers with shared encodings
From reading the documentation it seems like there's not an easy way to have multiple marks share the same encodings without repeating them. For example in [this example from the docs](https://vega.github.io/vega-lite/docs/layer.html#example):

```json
{
  "data": {"url": "data/stocks.csv"},
  "layer": [
    {
      "mark": "line",
      "encoding": {
        "x": {"field": "date", "type": "temporal"},
        "y": {"field": "price", "type": "quantitative"},
        "color": {"field": "symbol", "type": "nominal"}
      }
    },
    {
      "mark": "rule",
      "encoding": {
        "y": {
          "field": "price",
          "type": "quantitative",
          "aggregate": "mean"
        },
        "size": {"value": 2},
        "color": {"field": "symbol", "type": "nominal"}
      }
    }
  ]
}
```

the `color` scale is identical between the two layers, and most of the `y` scale is the same too. To avoid this kind of repetition it would be nice if the layers could "inherit" encodings from the top level, possibly augmenting them along the way:

```json
{
  "data": {"url": "data/stocks.csv"},
  "encoding": {
    "y": {"field": "price", "type": "quantitative"},
    "color": {"field": "symbol", "type": "nominal"}
  },
  "layer": [
    {
      "mark": "line",
      "encoding": {
        "x": {"field": "date", "type": "temporal"}
      }
    },
    {
      "mark": "rule",
      "encoding": {
        "y": { "aggregate": "mean" },
        "size": {"value": 2}
      }
    }
  ]
}
```

### PR DESCRIPTION
**Pull request by kanitw**
Fix #3384, Fix #3344

### CODE REVIEW

**Comment by kanitw** [comment]
@domoritz  -- The question is whether we should modify existing layer example to use this.

**Comment by domoritz** [comment]
Let's leave the examples as they are for now and wait for https://github.com/vega/vega-lite/issues/3516.

File: scripts/rename-schema.sh [hunk 1]
```diff
@@ -3,13 +3,14 @@
...
 perl -pi -e s,'GenericLayerSpec<CompositeUnitSpec>','LayerSpec',g build/vega-lite-schema.json
-perl -pi -e s,'GenericFacetSpec<CompositeUnitSpec>','FacetSpec',g build/vega-lite-schema.json
-perl -pi -e s,'GenericRepeatSpec<CompositeUnitSpec>','RepeatSpec',g build/vega-lite-schema.json
-perl -pi -e s,'GenericVConcatSpec<CompositeUnitSpec>','VConcatSpec',g build/vega-lite-schema.json
-perl -pi -e s,'GenericHConcatSpec<CompositeUnitSpec>','HConcatSpec',g build/vega-lite-schema.json
+perl -pi -e s,'GenericSpec<CompositeUnitSpec\,LayerSpec>','Spec',g build/vega-lite-schema.json
+perl -pi -e s,'GenericFacetSpec<CompositeUnitSpec\,LayerSpec>','FacetSpec',g build/vega-lite-schema.json
+perl -pi -e s,'GenericRepeatSpec<CompositeUnitSpec\,LayerSpec>','RepeatSpec',g build/vega-lite-schema.json
+perl -pi -e s,'GenericVConcatSpec<CompositeUnitSpec\,LayerSpec>','VConcatSpec',g build/vega-lite-schema.json
+perl -pi -e s,'GenericHConcatSpec<CompositeUnitSpec\,LayerSpec>','HConcatSpec',g build/vega-lite-schema.json
```
Comment (domoritz) [inline scripts/rename-schema.sh]: Use regex with replacer like here `perl -pi -e s,'TopLevel<(.*)>','TopLevel\1',g build/vega-lite-schema.json`

File: test/spec.test.ts [hunk 2]
```diff
@@ -26,7 +26,7 @@ describe('normalize()', function () {
         }
       };
       const config = initConfig(spec.config);
-      assert.deepEqual<GenericSpec<GenericUnitSpec<Encoding<Field>, string | MarkDef>>>(normalize(spec, config), {
+      assert.deepEqual<GenericSpec<any, any>>(normalize(spec, config), {
```
Comment (domoritz) [inline test/spec.test.ts]: Can you just delete this?

**Comment by kanitw** [comment]
> Let's leave the examples as they are for now and wait for #3516.

If you look at most layer example, the encodings usually have shared parts, but not exactly identical.  So #3516 won't help much.

**Comment by domoritz** [comment]
More minor comments above but looks good otherwise.

**Comment by domoritz** [comment]
We can merge https://github.com/vega/vega-lite/pull/3345 afterwards.

**Comment by kanitw** [comment]
Done.  I also add https://github.com/vega/vega-lite/pull/3515/commits/7cfeadaa38d2f2105c11c09b5c5b4dcd8da30d31 to update examples

**Pull request merged**